### PR TITLE
Route exported models by embedded metadata

### PR DESCRIPTION
**Pull request by glenn-jocher**
Training writes `best.pt`, so its exports (`best.onnx`, `best.engine`, `best.torchscript`, `best_openvino_model/`, ...) carry no task in their filename, and `guess_model_task` then falls back to `detect`: a segmentation export returns no masks, a pose export no keypoints, and an RT-DETR export is handed the YOLO detect predictor, which misreads its normalized `cxcywh` queries as pixel `xyxy`.

Every export already embeds the `Exporter` metadata (`task`, `head`, ...), so this reads it back before the task and model family are resolved, with the filename kept as the fallback for files that carry none:

- `read_export_metadata()` in `nn/backends/base.py` reads the TensorRT header, the TorchScript extra file, the TFLite zip entry, ONNX `metadata_props` (also for `_imx_model/` directories), CoreML `userDefined`, and the `metadata.yaml` sidecar of the directory formats, including the `_saved_model/` sibling of a frozen `.pb`. It reuses `onnx` and `coremltools`, which those backends already require, instead of a bespoke protobuf parser. MNN keeps its metadata in a flatbuffer `bizCode` and Triton serves it over HTTP, so those still resolve from the filename.
- `guess_model_task()` consults it first for path inputs, and `YOLO.__init__` uses its `head` to route exported RT-DETR models to `RTDETR`.
- The eleven backends that each duplicated the `metadata.yaml` sidecar lookup now call the shared reader, a net deletion.

```python
from ultralytics import YOLO

path = YOLO("yolo26n-seg.pt").export(format="onnx")  # rename to best.onnx
YOLO("best.onnx").predict("https://ultralytics.com/images/bus.jpg")  # task=segment, masks present, no warning
```

Verified locally with `yolo26n-seg.pt` and `yolo26n.pt` exported to ONNX and TorchScript and renamed to `best.*` (segment/detect resolve correctly via Python and CLI), reading a 94 MB ONNX takes 8 ms, and `tests/test_python.py::test_model_methods`, `tests/test_exports.py::test_export_onnx` and `::test_export_torchscript` pass.

Supersedes #25886. Thank you @artest08 for surfacing the problem and for the original implementation.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Exported models now resolve their task and model head from embedded metadata before using filename-based detection, allowing renamed exports such as `best.onnx` to load with the correct behavior.

### 📊 Key Changes

- Added `read_export_metadata()` in `ultralytics/nn/backends/base.py` to read metadata from TensorRT, TorchScript, TFLite, ONNX, IMX, CoreML, and directory-based export formats, including frozen TensorFlow graph sidecars.
- Updated `guess_model_task()` to prioritize embedded export task metadata and retain filename-based detection as a fallback.
- Updated `YOLO.__init__` to use the embedded `head` metadata and route exported `RTDETRDecoder` models to `RTDETR`.
- Replaced duplicated `metadata.yaml` loading in the export backends with the shared metadata reader.

### 🎯 Purpose & Impact

- Renamed segmentation, pose, and other exports can retain their correct task without relying on task-specific filenames.
- Exported RT-DETR models are routed to the RT-DETR implementation instead of the YOLO detection predictor.
- Backends continue to resolve metadata from existing sidecar files when embedded metadata is unavailable.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the shared metadata reader, task inference, RT-DETR routing, and backend sidecar integrations. Two realistic regressions remain: standalone `.pb` files can raise `StopIteration`, and ONNX task routing silently fails when only the runtime dependency is installed.

<details><summary>💬 Posted 2 inline comments</summary>

- 💡 **MEDIUM** `ultralytics/nn/backends/base.py:66` Task discovery now requires importing `onnx`, but this is only an optional `export-base` dependency in `pyproject.toml`; the ONNX inference backend itself checks for and uses `onnxruntime`. Because the broad exception below converts a missing `onnx` import into `{}`, an inference-only installation can load a renamed `best.onnx` but still silently route segmentation or pose exports as detect. Avoid making metadata-based task inference depend on the exporter package, or otherwise ensure this depe…
- 💡 **MEDIUM** `ultralytics/nn/backends/base.py:78` When a standalone frozen `.pb` has no sibling `<stem>_saved_model*/metadata.yaml`, `next(...)` raises `StopIteration` here. This is outside the `try`, and `guess_model_task()` calls this before `TensorFlowBackend` loads the graph, so `YOLO("model.pb")` now fails during construction instead of preserving the previous fallback behavior. Provide the existing `sidecar` as the default value.
</details>

**Review by UltralyticsAssistant — DISMISSED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed all 14 changed files, including metadata extraction, task inference, RT-DETR routing, and backend sidecar migration. The control flow is consistent with the existing exporter and backend contracts, with no evidence-backed correctness, compatibility, security, or performance issues. LGTM.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The previously reported standalone `.pb` failure is resolved by the default sidecar fallback now present on line 80. The shared reader, sidecar migrations, and RT-DETR routing are otherwise consistent; one remaining compatibility regression affects renamed segmentation/pose ONNX models used through OpenCV DNN without the optional `onnx` package.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/nn/backends/base.py:67` This makes task discovery depend on the optional `onnx` package even when the ONNX backend is using OpenCV DNN, whose load path only requires `cv2`. With `dnn=True`, an installation that has OpenCV but not `onnx` will fail the import, return `{}` through the broad exception below, and route a renamed segmentation or pose export as `detect`. Keep metadata extraction independent of the exporter-only dependency for the DNN path, or otherwise preserve embedded task metadata when `onnx` is unavailab…
</details>

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Review 4: the standalone `.pb` fallback remains fixed, and the shared sidecar migrations are consistent. The previously reported OpenCV DNN compatibility regression for renamed ONNX segmentation/pose exports is still unresolved; no new issues were found.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/nn/backends/base.py:67` This still makes task discovery depend on the optional `onnx` exporter package for the OpenCV DNN path. With `YOLO("best.onnx").predict(..., dnn=True)` in an installation that has OpenCV but not `onnx`, `check_requirements("onnx")` fails and the broad `except` returns `{}` before metadata can be used, so a renamed segmentation or pose export is routed as `detect` and uses the wrong predictor. Keep metadata discovery usable without the exporter-only dependency, or preserve the embedded metadata…
</details>

File: ultralytics/nn/backends/base.py [hunk 1]
```diff
@@ -38,6 +41,46 @@ def read_tflite_metadata(file: str | Path) -> dict | None:
...
+        if path.suffix == ".engine":  # length-prefixed JSON header
+            with open(path, "rb") as f:
+                return json.loads(f.read(int.from_bytes(f.read(4), "little", signed=True)))
+        if path.suffix == ".tflite":
+            return read_tflite_metadata(path) or {}
+        if path.suffix == ".torchscript":  # extra file written by torch.jit.save
+            with zipfile.ZipFile(path) as zf:
+                return json.loads(zf.read(next(n for n in zf.namelist() if n.endswith("extra/config.txt"))))
+        if path.suffix == ".onnx" or path.name.endswith("_imx_model"):  # IMX packages its ONNX in a directory
+            check_requirements("onnx")  # the ONNX backend requires it alongside onnxruntime
```
Comment (glenn-jocher) [inline ultralytics/nn/backends/base.py:67]: The OpenCV DNN path reads no metadata at all today (`cv2.dnn.readNetFromONNX` only), so with `dnn=True` and no `onnx` installed it resolves from the filename exactly as on main: no regression, and `check_requirements` auto-installs `onnx` by default so the common case now routes correctly too. Keeping this as is rather than adding a second ONNX reader for an edge case.

**Comment by glenn-jocher** [comment]
Closing — changes will be applied directly to #25886.

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 5

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Review 5: the standalone `.pb` fallback and shared sidecar migrations remain correct. The previously discussed OpenCV DNN/optional-`onnx` behavior is unchanged and has been addressed by the maintainer’s compatibility rationale. No new evidence-backed bugs, regressions, or performance issues found. LGTM.
